command/agent/exec/exec.go

```diff
@@ -0,0 +1,288 @@
+package exec
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/consul-template/child"
+	ctconfig "github.com/hashicorp/consul-template/config"
+	"github.com/hashicorp/consul-template/manager"
+	"github.com/hashicorp/go-hclog"
+	"go.uber.org/atomic"
+
+	"github.com/hashicorp/vault/command/agent/config"
+	"github.com/hashicorp/vault/command/agent/internal/ctmanager"
+	"github.com/hashicorp/vault/helper/useragent"
+	"github.com/hashicorp/vault/sdk/helper/pointerutil"
+)
+
+type ServerConfig struct {
+	Logger      hclog.Logger
+	AgentConfig *config.Config
+
+	Namespace string
+
+	// LogLevel is needed to set the internal Consul Template Runner's log level
+	// to match the log level of Vault Agent. The internal Runner creates it's own
+	// logger and can't be set externally or copied from the Template Server.
+	//
+	// LogWriter is needed to initialize Consul Template's internal logger to use
+	// the same io.Writer that Vault Agent itself is using.
+	LogLevel  hclog.Level
+	LogWriter io.Writer
+}
+
+type Server struct {
+	// config holds the ServerConfig used to create it. It's passed along in other
+	// methods
+	config *ServerConfig
+
+	// runner is the consul-template runner
+	runner *manager.Runner
+
+	// lookupMap is a list of templates indexed by their consul-template ID. This
+	// is used to ensure all Vault templates have been rendered before returning
+	// from the runner in the event we're using exit after auth.
+	lookupMap map[string][]*ctconfig.TemplateConfig
+
+	stopped *atomic.Bool
```

Comment:
should this be procStopped (similar to procStarted)?

Reply:
A better idea, just have an enum that represents the state of the child process:

```
const (
    processUnknown childProcessState = iota
    processStarted
    processStopped
    processExited
)
```
etc.

Reply:
I think a state enum makes more sense if we can reasonable update this in a concurrent safe matter.

Reply:
no, this is for the "server" part

not sure I agree with the enum, what would be the difference between `stopped` and `exited`?

Reply:
Those are just example states, we could drop `exited` and consolidate it into `processStopped` if it doesn't make sense. However, I do think having a single "state" variable would be more clear than having two booleans (started/stopped). With two variables, there is a danger of forgetting to update one, which would result in weird bugs later on.

Reply:
TODO: create enum
- not started
- started/running
- restarting
- stopped (for now this wouldn't be read by anyone but for possible future use)